### PR TITLE
Refactor handleKeydown Method for Improved Readability and Maintainability

### LINKED ISSUE #795: [KTable]: Refactoring the handleKeydown method in KTable
**🌱 Are you new to the codebase? Welcome! Please see [the contributing guidelines](https://github.com/learningequality/kolibri-design-system/blob/develop/CONTRIBUTING.md).**
## Summary
In #727, we introduced a new component—`KTable`—which provides a flexible, accessible table layout with enhanced keyboard navigation capabilities. As part of this, the `handleKeydown` method is responsible for managing keyboard navigation between table cells, handling focus, and managing the arrow, tab, and other keys.
The current implementation of the `handleKeydown` method has grown more complex over time, making it harder to maintain and reason about. The goal of this issue is to refactor the method into a more readable and maintainable format, while preserving the functionality and accessibility of the keyboard navigation.
## The Value Add 
- Improving overall code quality in the `KTable` component.
- Ensuring that future modifications to the keyboard navigation behavior can be made more easily.
- Improving overall code quality in the KTable component.
## Possible Tradeoffs
Potential performance considerations: although the goal is to refactor for readability, care should be taken to avoid performance regressions, especially when handling key events frequently during table navigation.
## Guidance
Please refer to [this discussion](https://github.com/learningequality/kolibri-design-system/pull/727#discussion_r1725156184) for a better understanding of the implementation. In Summary,
- Focus on breaking down the handleKeydown logic into smaller, modular methods that perform specific actions, such as handling different key events (i.e., arrow keys, tab, enter).
- Avoid introducing unnecessary complexity; ensure the refactor simplifies the code.
- Ensure that the refactor maintains the existing functionality, including:
  - Navigating between cells using arrow keys.
  - Handling focus management across rows and columns.
  - Supporting sticky headers and columns as described in the component behavior.
- Include unit tests for the refactored code, especially around the key navigation and focus management aspects.
## Acceptance Criteria 

- [ ] No regressions in existing keyboard navigation functionality.
- [ ] Code is significantly more readable and easier to understand.
- [ ] Unit tests are added or updated to cover the refactored method (if any)
- [ ] No performance regressions due to the refactor.

### PR DESCRIPTION
**Pull request by shivam-daksh**
## Description

This PR addresses the issue described in [Issue #795](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.esm.html). The `handleKeydown` method in the `KTable` component has been refactored to improve readability and maintainability while preserving the existing functionality and accessibility of the keyboard navigation.
#### Issue addressed

#795

Addresses #*PR# HERE*

### Summary

In [#727](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.esm.html), we introduced a new component—KTable—which provides a flexible, accessible table layout with enhanced keyboard navigation capabilities. As part of this, the handleKeydown method is responsible for managing keyboard navigation between table cells, handling focus, and managing the arrow, tab, and other keys.

The current implementation of the handleKeydown method has grown more complex over time, making it harder to maintain and reason about. The goal of this issue is to refactor the method into a more readable and maintainable format while preserving the functionality and accessibility of the keyboard navigation.

### Changes Made

1. Refactored handleKeydown Method:
-- Broke down the logic into smaller, modular methods for handling different key events.
-- Ensured that the Tab key mimics the functionality of the ArrowRight key.
-- Ensured that the Shift+Tab key mimics the functionality of the ArrowLeft key and moves to the last cell when the first cell is focused.
```
methods: {
  handleKeydown(event, rowIndex, colIndex) {
    const key = event.key;
    const totalRows = this.rows.length;
    const totalCols = this.headers.length;

    let nextRowIndex = rowIndex;
    let nextColIndex = colIndex;

    switch (key) {
      case 'ArrowUp':
        ({ nextRowIndex, nextColIndex } = this.handleArrowUp(rowIndex, colIndex, totalRows));
        break;
      case 'ArrowDown':
        ({ nextRowIndex, nextColIndex } = this.handleArrowDown(rowIndex, colIndex, totalRows));
        break;
      case 'ArrowLeft':
        ({ nextRowIndex, nextColIndex } = this.handleArrowLeft(rowIndex, colIndex, totalRows, totalCols));
        break;
      case 'ArrowRight':
        ({ nextRowIndex, nextColIndex } = this.handleArrowRight(rowIndex, colIndex, totalRows, totalCols));
        break;
      case 'Enter':
        this.handleEnter(rowIndex, colIndex);
        break;
      case 'Tab':
        ({ nextRowIndex, nextColIndex } = this.handleTabKey(event, rowIndex, colIndex));
        break;
      default:
        return;
    }

    this.updateFocus(nextRowIndex, nextColIndex);
    event.preventDefault();
  },

  handleArrowUp(rowIndex, colIndex, totalRows) {
    let nextRowIndex = rowIndex === -1 ? totalRows - 1 : rowIndex - 1;
    return { nextRowIndex, nextColIndex: colIndex };
  },

  handleArrowDown(rowIndex, colIndex, totalRows) {
    let nextRowIndex = rowIndex === -1 ? 0 : (rowIndex + 1) % totalRows;
    return { nextRowIndex, nextColIndex: colIndex };
  },

  handleArrowLeft(rowIndex, colIndex, totalRows, totalCols) {
    let nextRowIndex = rowIndex;
    let nextColIndex = colIndex;

    if (rowIndex === -1) {
      if (colIndex > 0) {
        nextColIndex = colIndex - 1;
      } else {
        nextColIndex = totalCols - 1;
        nextRowIndex = totalRows - 1;
      }
    } else if (colIndex > 0) {
      nextColIndex = colIndex - 1;
    } else {
      nextColIndex = totalCols - 1;
      nextRowIndex = rowIndex > 0 ? rowIndex - 1 : -1;
    }

    return { nextRowIndex, nextColIndex };
  },

  handleArrowRight(rowIndex, colIndex, totalRows, totalCols) {
    let nextRowIndex = rowIndex;
    let nextColIndex = colIndex;

    if (colIndex === totalCols - 1) {
      if (rowIndex === totalRows - 1) {
        nextColIndex = 0;
        nextRowIndex = -1;
      } else {
        nextColIndex = 0;
        nextRowIndex = rowIndex + 1;
      }
    } else {
      nextColIndex = colIndex + 1;
    }

    return { nextRowIndex, nextColIndex };
  },

  handleEnter(rowIndex, colIndex) {
    if (rowIndex === -1 && this.sortable) {
      this.handleSort(colIndex);
    }
  },

  handleTabKey(event, rowIndex, colIndex) {
    const currentCell = this.getCell(rowIndex, colIndex);

    if (!currentCell) return false;

    const focusableElements = this.getFocusableElements(currentCell);
    const focusedElementIndex = focusableElements.indexOf(document.activeElement);

    if (focusableElements.length > 0) {
      if (!event.shiftKey) {
        if (focusedElementIndex < focusableElements.length - 1) {
          focusableElements[focusedElementIndex + 1].focus();
          event.preventDefault();
          return { nextRowIndex: rowIndex, nextColIndex: colIndex };
        }
      } else {
        if (focusedElementIndex > 0) {
          focusableElements[focusedElementIndex - 1].focus();
          event.preventDefault();
          return { nextRowIndex: rowIndex, nextColIndex: colIndex };
        }
      }
    }

    const totalRows = this.rows.length;
    const totalCols = this.headers.length;
    let nextRowIndex = rowIndex;
    let nextColIndex = colIndex;

    if (!event.shiftKey) {
      // Mimic ArrowRight key
      if (colIndex === totalCols - 1) {
        nextColIndex = 0;
        nextRowIndex = rowIndex === totalRows - 1 ? -1 : rowIndex + 1;
      } else {
        nextColIndex = colIndex + 1;
      }
    } else {
      // Mimic ArrowLeft key
      if (colIndex === 0) {
        if (rowIndex === 0) {
          // Move to the last cell if the first cell is focused
          nextColIndex = totalCols - 1;
          nextRowIndex = totalRows - 1;
        } else {
          nextColIndex = totalCols - 1;
          nextRowIndex = rowIndex > 0 ? rowIndex - 1 : -1;
        }
      } else {
        nextColIndex = colIndex - 1;
      }
    }

    return { nextRowIndex, nextColIndex };
  },

  moveToNextCell(rowIndex, colIndex) {
    const totalRows = this.rows.length;
    const totalCols = this.headers.length;

    let nextRowIndex = rowIndex;
    let nextColIndex = colIndex;

    if (colIndex < totalCols - 1) {
      nextColIndex = colIndex + 1;
    } else if (rowIndex < totalRows - 1) {
      nextColIndex = 0;
      nextRowIndex = rowIndex + 1;
    }

    this.focusCell(nextRowIndex, nextColIndex);
  },

  moveToPreviousCell(rowIndex, colIndex) {
    const totalRows = this.rows.length;
    const totalCols = this.headers.length;

    let nextRowIndex = rowIndex;
    let nextColIndex = colIndex;

    if (colIndex > 0) {
      nextColIndex = colIndex - 1;
    } else if (rowIndex > 0) {
      nextColIndex = totalCols - 1;
      nextRowIndex = rowIndex - 1;
    }

    this.focusCell(nextRowIndex, nextColIndex);
  },

  updateFocus(nextRowIndex, nextColIndex) {
    this.focusCell(nextRowIndex, nextColIndex);
    this.focusedRowIndex = nextRowIndex === -1 ? null : nextRowIndex;
    this.focusedColIndex = nextColIndex;
    this.highlightHeader(nextColIndex);
  }
}
```
This pull request introduces several enhancements and refactorings to the keyboard navigation and focus management logic in the `KTable` component. The changes improve the handling of various keyboard events and modularize the code for better readability and maintainability.

### Enhancements to keyboard navigation:

* **Enter and Tab Key Handling**: Added logic to handle `Enter` key for sorting columns and `Tab` key for navigating between focusable elements within a cell. (`lib/KTable/index.vue`)
* **Arrow Key Navigation**: Refactored the logic for `ArrowUp`, `ArrowDown`, `ArrowLeft`, and `ArrowRight` keys to simplify and enhance navigation between cells. (`lib/KTable/index.vue`)

### Code modularization:

* **Extracted Methods**: Created `getNextIndexes`, `handleTabKey`, `moveToNextCell`, `moveToPreviousCell`, and `getFocusableElements` methods to modularize and clean up the main event handler logic. (`lib/KTable/index.vue`)
## Changelog

  - **Description:** Refactored `handleKeydown` method into smaller, modular methods for handling different key events. Ensured that the `Tab` key mimics the functionality of the `ArrowRight` key, and the `Shift+Tab` key mimics the functionality of the `ArrowLeft` key. \\
  - **Products impact:** none
  - **Addresses:** #795 
  - **Components:** `KTable `
  - **Breaking:** no
  - **Impacts a11y:** yes
  - **Guidance:** No special guidance required for consumers.

## Steps to test

1. Step 1
2. Step 2
3. ...

## Testing checklist

- [ ] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical and brittle code paths are covered by unit tests
- [ ] The change is described in the changelog section above

## Reviewer guidance

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

### CODE REVIEW

**Comment by shivam-daksh** [comment]
Hi @MisRob & @BabyElias, kindly review it. 
I've made slight changes in the functionality of **tab** and **shift+tab** key. Basically, these are mimicing the functionality of left arrow and right arrow key respectively.

**Comment by BabyElias** [comment]
@shivam-daksh Can you please update the Changelog in description with all the fields that were originally in the template? 
You can see the Changelog [here](https://github.com/learningequality/kolibri-design-system/pull/798) for reference. 

Also, please run the `yarn lint fix` command to fix linting issues with the PR. Thanks :)

**Comment by shivam-daksh** [comment]
@BabyElias, I've updated the changelog section in description. Please have a look at it.
 I ran `yarn lint fix` command but it didn't made any changes to the code formatting.

**Comment by BabyElias** [comment]
@shivam-daksh The Changelog looks good 👍

For the linter -  maybe try `yarn install --force` then `yarn lint-fix` again, perhaps it is an issue with your currently installed version of kolibri-tools (where the linter lives). However, from the issues that I can see with the failed linter test, it is caused due to declaration of unused variables and the linter can't fix these issues on its own (it can fix alignment on its own, but has issues with code refactor related issues like these). So, this change needs to be manually done by you for the tests to pass. You can have a look at the issues after execution of `yarn lint-fix` in the console itself. Hope this helps .

**Comment by shivam-daksh** [comment]
Hey @BabyElias, I've just passed the linting test by commenting the unsued code. Could you please review it now?

**Comment by BabyElias** [comment]
Hey @shivam-daksh , good work overall :) 
I had a few pointers related to keyboard nav features that existed earlier and have now been affected by the refactor :
- Shift+Tab key combination does not focus on focusable elements within the cell (while tab key does)
- Using arrow keys, the corresponding cell header and entire cell get highlighted (as expected). The same is supposed to happen when the tab and Shift+Tab keys are used, but that does not happen.

Don't know if you have an idea about the playground page yet to test your changes or not, you can have a look at it [here](https://github.com/learningequality/kolibri-design-system/blob/develop/dev_docs/03_how_to_update_library.md#1-preview-updates-in-the-kolibri-design-system-environment). This comes in extremely handy to preview the output for code changes as and when they are done.
I would highly recommend testing your keyboard navigation features by copy-pasting [this](https://github.com/learningequality/kolibri-design-system/blob/eb923d275021f1d0ba3b97c9cd59a2f6165dbc0c/docs/pages/playground.vue) playground page in your local repository. This playground page contains example for the table with focusable elements as well, that can help you better in reviewing the code.
Remember to not commit playground page changes, just keep them for local reference and testing. Thanks!

**Comment by shivam-daksh** [comment]
Hi @BabyElias , `Tab` and `Shift+Tab` key is working as expected. Please take a look.

**Comment by BabyElias** [comment]
Hey @shivam-daksh :)
The row and header highlight work just perfect now, kudos on that!
A few issues that I can observe here though,
- Using tab key, we are able to focus on the focusable elements within the cell but using Shift+tab key while moving backwards it the table, we cannot. This needs to be fixed.
-  The observation, that Tab Key and Shift+Tab key mimic left and right arrow keys is valid when navigating inside the table, however, Tab and Shift+Tab key are originally meant to move from one element to other according to the [DOM order](https://accessibleweb.com/question-answer/what-is-a-dom/). This means, that once a user reaches the end of the table (bottom rightmost cell), the tab key should function as it's default and move out of the table to the next element within DOM order (unlike arrow keys that keep moving within the table). Similarly, for Shift+Tab key in the opposite order. This doesn't need to be engineered manually as such- you just have to enable default tab behaviour when reaching end of table, you can see how it was implemented in the previous code and take inspiration accordingly. 

Thanks!

**Comment by shivam-daksh** [comment]
Hey @BabyElias , Please review it again.

**Comment by BabyElias** [comment]
@shivam-daksh The Tab key behaviour is as expected now, great job there!
I am still unable to focus elements using Shift+Tab key combo, not sure if you have fixed that or are yet to do the same.

Also, I was having a look at the code, and I could see that a number of comments that existed earlier have been removed by the new commits. We try to keep our code as well documented and easy to understand as possible so that it can be easier for future contributors to make sense of the code-base and contribute in a better way, especially with something as complicated as the `handlekeydown `method which might not be understandable for everyone. So, it would be really helpful if you could add a few of those comments back in the further commits or maybe write something of your own to explain the newly done changes :)

**Comment by shivam-daksh** [comment]
Hi @BabyElias, Thanks for the reminding! I’ve re-added the lost comments during the refactoring to help clarify the code for future contributors.

Regarding the Shift+Tab functionality, I’m a bit unclear on how it should be working differently. As far as I can tell, its behavior remains the same as before the refactor.  If you could provide a little visual help or specific examples of what you expect, that would really help me address the issue effectively!

Looking forward to your input!


https://github.com/user-attachments/assets/4ddc11a1-e939-4e0f-bb63-c9dca018cc23

**Comment by BabyElias** [comment]
@shivam-daksh,
About the Shift+Tab key, the KTable example in the previous video does not contain any focusable elements within the cell (like a button, the example for such a table can be found in the playground page I referenced in an earlier comment). So, if you use that example, then the way Tab key works in the forward navigation - Shift+Tab key is expected to work in the same way in backward direction. 
For eg - if we have three cells in a row in which third one has a button, the order of navigation moves as 
Tab key  : cell1->cell2->cell3->button in cell3->next row, first cell
Shift+Tab key : next row,first cell -> button in cell3->cell3->cell2->cell1
Basically, the button should be focussed using both tab key and shift+tab keys.
Hope this makes it a little clearer than before, in case you still find any difficulty in understanding - I'll also record a video demo for the same. Currently, have some technical issues with doing that so :|

**Comment by shivam-daksh** [comment]
@BabyElias, Okay. I'm sorry. Totally misunderstood that 😆 .
I'm on it.

**Comment by MisRob** [comment]
Thanks @shivam-daksh for this work and @BabyElias for taking care of the review, we really appreciate it :)

**Comment by MisRob** [comment]
>  Okay. I'm sorry. Totally misunderstood that 😆 .

No worries, @shivam-daksh. The table has many features that are not immediately obvious, we're currently discussing how to document technical requirements, including a11y, in detail, to guide future work on more complex components. Meanwhile, fortunately @BabyElias and @radinamatic have everything in their mind.

**Comment by MisRob** [comment]
Hi @shivam-daksh, no time pressure around this task, just checking in if there's anything you'd need since it seems some aspects are a bit tricky.

I also noted @BabyElias's

> Hope this makes it a little clearer than before, in case you still find any difficulty in understanding - I'll also record a video demo for the same. Currently, have some technical issues with doing that so :|

and if that'd help, I could try to do the recording for you

**Comment by shivam-daksh** [comment]
@BabyElias and @MisRob, I am really sorry for extending that long. My exams are going on. However, I will finish it at end of this week. 
Also, I understand the problem and have figured out how to properly refactor it. Thanks to @BabyElias for the explaination.

**Comment by MisRob** [comment]
> @BabyElias and @MisRob, I am really sorry for extending that long. My exams are going on. However, I will finish it at end of this week.
Also, I understand the problem and have figured out how to properly refactor it. Thanks to @BabyElias for the explaination.

@shivam-daksh Really please take all the time you need - it's absolutely fine for volunteers to work on tasks for longer periods of time - weeks, even months. Most of our 'help wanted' are intentionally not time sensitive, and for those that are, we'd inform volunteers beforehand.

You're generously volunteering your time and energy and there's really no need to make it into feeling bad! I hope you can focus on your exams and don't forget to get plenty of rest ;)

I'm glad to hear that the expectations are clearer. Thanks @BabyElias for all the guidance for @shivam-daksh.

**Comment by shivam-daksh** [comment]
Hey @BabyElias, please have a look at this.

**Comment by BabyElias** [comment]
Hey @shivam-daksh, so happy to see all the pieces finally put together!
The functionality is almost completely implemented now, I'll make a few changes and add them as review here so that you can also understand the few parts that remain. You can just commit them and the code will be good to go :)

File: lib/KTable/index.vue [hunk 1]
```diff
@@ -294,156 +294,146 @@
...
+        this.focusedRowIndex = nextRowIndex === -1 ? null : nextRowIndex;
+        this.focusedColIndex = nextColIndex;
+        this.highlightHeader(nextColIndex);
+        event.preventDefault();
+      },
 
-            // Collect focusable elements using native DOM methods
-            const focusableElements = [];
+      handleEnterKey(colIndex) {
+        if (this.sortable) {
```
Comment (BabyElias) [inline lib/KTable/index.vue]: 01 - 
The Enter key should only sort the column if it is pressed on the header cells (i.e `rowIndex=== -1 && this.sortable`)
So, both rowIndex and colIndex will have to be passed as props to `handleEnterKey`

File: lib/KTable/index.vue [hunk 2]
```diff
@@ -294,156 +294,146 @@
...
-        const key = event.key;
+        switch (event.key) {
+          case 'ArrowUp':
+          case 'ArrowDown':
+          case 'ArrowLeft':
+          case 'ArrowRight':
+            this.handleArrowKeys(event.key, rowIndex, colIndex);
+            break;
+          case 'Enter':
+            this.handleEnterKey(colIndex);
```
Comment (BabyElias) [inline lib/KTable/index.vue]: Based on point 01 about Enter Key, both colIndex and rowIndex need to be passed to handleEnter key since sorting should only be executed when pressed on header cells

File: lib/KTable/index.vue [hunk 3]
```diff
@@ -294,156 +294,146 @@
...
+            const prevCell = this.getCell(nextRowIndex, nextColIndex);
+            const prevFocusableElements = this.getFocusableElements(prevCell);
+            const prevCellAndFocusableElements = [prevCell, ...prevFocusableElements];
+            prevCellAndFocusableElements[prevCellAndFocusableElements.length - 1].focus();
+            this.updateFocusState(nextRowIndex, nextColIndex);
+            event.preventDefault();
+          }
         }
+      },
+      updateFocusState(rowIndex, colIndex) {
```
Comment (BabyElias) [inline lib/KTable/index.vue]: Adding this function makes the code look much cleaner :)

File: lib/KTable/index.vue [hunk 4]
```diff
@@ -294,156 +294,146 @@
...
-            }
-            break;
-          case 'Tab': {
-            // Identify all focusable elements inside the current cell
-            const currentCell = this.getCell(rowIndex, colIndex);
+        }
+        this.focusCell(nextRowIndex, nextColIndex);
+        this.focusedRowIndex = nextRowIndex === -1 ? null : nextRowIndex;
+        this.focusedColIndex = nextColIndex;
+        this.highlightHeader(nextColIndex);
```
Comment (BabyElias) [inline lib/KTable/index.vue]: Since we already have an updateFocusState function for the following tasks, I think it will be better to use that here too.

File: lib/KTable/index.vue [hunk 5]
```diff
@@ -294,156 +294,146 @@
...
+            this.updateFocusState(nextRowIndex, nextColIndex);
+            event.preventDefault();
+          }
         }
+      },
+      updateFocusState(rowIndex, colIndex) {
+        this.focusedRowIndex = rowIndex === -1 ? null : rowIndex;
+        this.focusedColIndex = colIndex;
+        this.highlightHeader(colIndex);
+      },
```
Comment (BabyElias) [inline lib/KTable/index.vue:490]: If you have a look at the previous code,a function has been left altogether in the code refactor - [`focusCell`](https://github.com/BabyElias/kolibri-design-system/blob/c04026a44fd4646a7a4f5763560b1d8ee1f6e2ac/lib/KTable/index.vue#L439). This function is responsible for scrolling cells into the viewport when navigating using the keyboard and ensures smooth accessible experience throughout the table. We can add this function in the `updateFocusState` function itself as
 `this.focusCell(rowIndex,colIndex);` on line 429 and it should work just fine.

File: lib/KTable/index.vue [hunk 6]
```diff
@@ -294,156 +294,146 @@
...
+            event.preventDefault();
           }
-
-          default:
+        } else {
+          // Shift+Tab key navigation
+          if (focusedIndex > 0) {
+            // Move to the previous focusable element within the cell
+            cellAndFocusableElements[focusedIndex - 1].focus();
+            event.preventDefault();
```
Comment (BabyElias) [inline lib/KTable/index.vue:468]: This condition would need to be changed to -

```
 if (focusedIndex < cellAndFocusableElements.length - 1) {
            // Move to the previous focusable element within the cell
            cellAndFocusableElements[focusedIndex + 1].focus();
            event.preventDefault();
            return;
          } 

```
(Same as when navigating using tab key, since there is some ambiguity caused with the focusCell function when the current condition is used)
You can have a look at it yourself - like if you don't change the shift+tab key condition here with the tab key condition and add `this.focusCell` to `updateFocusState`, the focusable elements within cells would no longer be focused when we navigate using Shift+Tab keys.

**Review by BabyElias — CHANGES_REQUESTED** [review]

File: lib/KTable/index.vue (hunk 6 above)
Comment (shivam-daksh) [reply lib/KTable/index.vue:468]: @BabyElias Yes, I did it myself and it was having problem with `this.focusCell` as you said. Now it's fixed and working fine as expected.

I have a bit of confusion regarding the navigation order when using `shift+tab` to move back to a previous cell and the focusable elements within that cell.

For instance, given the following layout:

- row n: cell1, cell2, cell3 > (button1, button2)
- row n+1: cell1, cell2, cell3 > (button1, button2)

If the focus is on `cell1` of row[n+1] and `shift+tab` is pressed, should the focus move to:
1. `cell3` of row[n]?
2. `button1` or `button2` within `cell3` of row[n]?
Or should it follow the sequence: `cell3` → `button1` → `button2` and so on? (this is our current implementation)

Another example:
- row1: cell1, cell2, cell3 > (button1, button2)
- row2: cell1, cell2, cell3 > (button1, button2)

Suppose the focus is on `cell3` of row1 (which contains focusable elements) and we press `shift+tab`. Should the focus move to `cell2` or `button1`/`button2` ? Personally, I suggest that in this case, the focus should move to `cell2` instead of `button1`, as moving directly to the button could lead to a slightly awkward user experience when navigating backward using `shift+tab`.

For now, I’ve implemented it as you suggested. Please let me know if there’s anything wrong with it. :)

Comment (BabyElias) [reply lib/KTable/index.vue:468]: @shivam-daksh For the first layout, the cell should follow `cell3 → button1 → button2 `, as in the current implementation - as a user should only be able to move to the focusable elements when tab key is pressed, otherwise tab and arrow keys should work the same way.
For the second layout, ideally it should be `cell2→button1/button2` when using Shift+Tab key - which is not how it currently works in our implementation. This is something which I had tried figuring out earlier as well but eventually it just lost direction. If you are able to implement this expected nav behaviour with `focusCell` functionality intact - that would be just perfect. No pressure around it, in case you wish to leave it as it is. 
Thank you for your time and efforts!

Comment (shivam-daksh) [reply lib/KTable/index.vue:468]: > ideally it should be `cell2→button1/button2` when using Shift+Tab key - which is not how it currently works in our implementation.

I'm in bit confusion. Do you mean if we are `cell3` of row1 then pressing `shift+tab` should be move to `cell2` or `button1/button2`? Because what I'm infering from what you wrote that line is that if our current focus is at `cell3` then pressing `shift+tab` should move to `cell2` then `button1/button2` of `cell3`.

Comment (BabyElias) [reply lib/KTable/index.vue:468]: Oh, my bad! Let me reiterate,
If we are on `cell3 of row1`, it should go to `cell2 of row1`. If we are on cell1 of row2, then pressing Shift+Tab we should move as
`row2,cell1->row1,cell3,button1/button2->row1,cell3->row1,cell2`

What I talked about in the previous comment was movement like this - 
`row2,cell1->row1,cell3->row1,cell3,button1/button2->row1,cell2`, but now I realise this is illogical - as the purpose of Shift+Tab key is to move in the exact reverse order as Tab key (opposite in DOM order) and so the movement from current cell to prev cell button to prev cell is most ideal. Apologies for any confusion caused, hope this makes things clear.

Comment (shivam-daksh) [reply lib/KTable/index.vue:468]: >as the purpose of Shift+Tab key is to move in the exact reverse order as Tab key (opposite in DOM order) and so the movement from current cell to prev cell button to prev cell is most ideal. 

Thanks for getting it. That is exactly 💯  what I was trying to say. 
So, should I go for it now (which may take couple of days or more)? or Should we leave it as it for now and open a new issue which can be solved later anytime?

**Comment by MisRob** [comment]
Wonderful work here @shivam-daksh and @BabyElias.

As agreed with @BabyElias, @AlexVelezLl will join us here to offer second eyes for the code as soon as all @BabyElias's feedback is resolved. I will then prepare test environment for our QA team in Kolibri.

File: lib/KTable/index.vue (hunk 6 above)
Comment (BabyElias) [reply lib/KTable/index.vue:468]: In case you are willing to do it yourself, doing it in this PR itself irrespective of the time works! Otherwise we can also open up a follow-up issue too . Anything that suits you.

Comment (MisRob) [reply lib/KTable/index.vue:468]: I am fine leaving you to decide, just a recommendation - usually whenever follow-up issue is possible, it's generally good practice. Quite often some unforeseen things come up that potentially block work that otherwise could be merged, and also reviews and QA are easier if it can be done in smaller chunks.

One way or another, we're happy to have you do that update @shivam-daksh :)

Comment (shivam-daksh) [reply lib/KTable/index.vue:468]: @MisRob and @BabyElias , I will prefer opening a follow-up issue. 😊

Comment (MisRob) [reply lib/KTable/index.vue:468]: Lovely @shivam-daksh , feel free to go ahead and if you can then comment there right away, we can assign you for future if it's your wish to have it reserved. It's very valuable however not pressing at all, so you can return to it any time you wish.

File: lib/KTable/index.vue [hunk 7]
```diff
@@ -294,156 +294,144 @@
...
             break;
           case 'ArrowDown':
-            if (rowIndex === -1) {
-              nextRowIndex = 0;
-            } else if (rowIndex === totalRows - 1) {
-              nextRowIndex = -1;
-            } else {
-              nextRowIndex = (rowIndex + 1) % totalRows;
-            }
+            nextRowIndex = rowIndex === -1 ? 0 : rowIndex === lastRowIndex ? -1 : rowIndex + 1;
```
Comment (AlexVelezLl) [inline lib/KTable/index.vue]: Lets avoid nested ternaries, this makes the code much harder to read.

File: lib/KTable/index.vue [hunk 8]
```diff
@@ -294,156 +294,144 @@
...
-              }
-            } else if (colIndex > 0) {
-              nextColIndex = colIndex - 1;
+              nextColIndex = colIndex > 0 ? colIndex - 1 : lastColIndex;
+              nextRowIndex = colIndex === 0 ? lastRowIndex : -1;
             } else {
-              nextColIndex = totalCols - 1;
-              nextRowIndex = rowIndex > 0 ? rowIndex - 1 : -1;
+              nextColIndex = colIndex > 0 ? colIndex - 1 : lastColIndex;
+              nextRowIndex = colIndex === 0 ? (rowIndex > 0 ? rowIndex - 1 : -1) : rowIndex;
```
Comment (AlexVelezLl) [inline lib/KTable/index.vue]: Idem, lets avoid these nested ternaries.

File: lib/KTable/index.vue [hunk 9]
```diff
@@ -294,156 +294,144 @@
...
-              }
+        if (!event.shiftKey) {
+          // Tab key navigation
+          if (focusedIndex < cellAndFocusableElements.length - 1) {
+            // Move to the next focusable element within the cell
+            cellAndFocusableElements[focusedIndex + 1].focus();
+            event.preventDefault();
+            return;
+          } else {
+            // Move to the first focusable element of the next cell
```
Comment (AlexVelezLl) [inline lib/KTable/index.vue]: "Move to the next cell"

File: lib/KTable/index.vue [hunk 10]
```diff
@@ -294,156 +294,144 @@
...
-                }
-              }
+              return; // Allow default Tab behavior when reaching the end
             }
-
-            break;
+            const nextCell = this.getCell(nextRowIndex, nextColIndex);
+            const nextFocusableElements = this.getFocusableElements(nextCell);
+            const nextCellAndFocusableElements = [nextCell, ...nextFocusableElements];
+            nextCellAndFocusableElements[0].focus();
```
Comment (AlexVelezLl) [inline lib/KTable/index.vue]: Why do we need this `nextCellAndFocusableElements` if we always will be just focusing the index 0? We dont do anything with this array after that. That means that we will be always focusing the nextCell. We dont need to query the focusable elments of the next cell.

File: lib/KTable/index.vue [hunk 11]
```diff
@@ -294,156 +294,144 @@
...
+              nextColIndex = totalCols - 1;
+              nextRowIndex = rowIndex - 1;
+            } else {
+              return; // Allow default Shift+Tab behavior when at the beginning
+            }
+            const prevCell = this.getCell(nextRowIndex, nextColIndex);
+            const prevFocusableElements = this.getFocusableElements(prevCell);
+            const prevCellAndFocusableElements = [prevCell, ...prevFocusableElements];
+            prevCellAndFocusableElements[prevCellAndFocusableElements.length - 1].focus();
+            this.updateFocusState(nextRowIndex, nextColIndex);
```
Comment (AlexVelezLl) [inline lib/KTable/index.vue:480]: This "update focus state" is already taking care of the cell focus, and we are just focusing the prevCell, and not taking into account the "prevCellAndFocusableElements" anyways. We should keep this method, without trying to focus the last focusable element of the previous cell. That means, removing these lines:

```javascript
            const prevCell = this.getCell(nextRowIndex, nextColIndex);
            const prevFocusableElements = this.getFocusableElements(prevCell);
            const prevCellAndFocusableElements = [prevCell, ...prevFocusableElements];
            prevCellAndFocusableElements[prevCellAndFocusableElements.length - 1].focus();
```

File: lib/KTable/index.vue (hunk 10 above)
Comment (AlexVelezLl) [reply lib/KTable/index.vue]: Btw in the `updateFocusState` we are already focusing the next cell, so there is no need of any of these lines:

```javascript
            const nextCell = this.getCell(nextRowIndex, nextColIndex);
            const nextFocusableElements = this.getFocusableElements(nextCell);
            const nextCellAndFocusableElements = [nextCell, ...nextFocusableElements];
            nextCellAndFocusableElements[0].focus();
```

**Review by AlexVelezLl — DISMISSED** [review]
Thank you @shivam-daksh! This is looking good. I have left some minor comments on small things I noticed. Please let me know if you have any questions :)

**Comment by AlexVelezLl** [comment]
Thank you @shivam-daksh! Code changes looks good to me! I will not yet approve this PR since @MisRob will check this with our QA team, but we can have my review as :heavy_check_mark:. I noticed that the issue description mentioned that we needed to add unit tests to this part of the code. I am not sure if you already discussed about this with @BabyElias or @MisRob (I didnt read the whole discussion). But pinging them in case the unit tests are required c:. Thank you!

**Comment by BabyElias** [comment]
The code changes do not introduce any such new functionalities that might need to be covered by new unit tests. So, seems good to go from my end too :)

**Comment by MisRob** [comment]
Thanks all of you, I will get ready the test Kolibri PR for QA

**Comment by MisRob** [comment]
@shivam-daksh would you please merge the latest `develop` to this branch? (this will pull in https://github.com/learningequality/kolibri-design-system/pull/824 I've just merged and we will be sure we have the very latest `KTable` before QA). I could also do it later on if you don't mind I pushed to your branch.

**Comment by shivam-daksh** [comment]
@MisRob, I've merged the develop into our current branch. The code is runnnig fine and table is working properly but having a warning in terminal. This happened after merge.

js
```
 ERROR  [Vue warn]: Invalid prop: custom validator check failed for prop "headers".                                  10:59:01

found in

---> <KTable> at lib/KTable/index.vue
       <Playground> at docs/pages/playground/index.vue
         <Nuxt>
           <.nuxt/layouts/default.vue> at .nuxt/layouts/default.vue
             <Root>
```

**Comment by BabyElias** [comment]
That is just a validator warning since a [new PR](https://github.com/learningequality/kolibri-design-system/pull/824) merged for KTable mandates having a columnID for each header row. Nothing blocking.

**Comment by MisRob** [comment]
Test PR in Kolibri open here https://github.com/learningequality/kolibri/pull/12914. If QA confirms no regressions, we can merge this work.

**Comment by MisRob** [comment]
Hi @shivam-daksh, I'm sorry we're behind. Just wanted to let you know it's still in our testing queue and we keep an eye on this work.

**Comment by MisRob** [comment]
QA passed so we can merge. Thank you @shivam-daksh!

**Review by MisRob — APPROVED** [review]

**Comment by shivam-daksh** [comment]
Thanks @MisRob, Finally, it is merged! 😄
